### ISSUE TITLE
Pointer#copy_from / Pointer#move_from performance regression

### ISSUE THREAD

**Issue by ssvb** [manastech/crystal#572]
After the #568 crash has been fixed, we have a performance regression (especially in the non-optimized build). The following program can be used as a benchmark:

```ruby
require "benchmark"
Benchmark.bm { |x|
  x.report("shift") {
    a = [0] ; 80000.times { |i| a.push(i) }
    # This is unreasonably slow in Crystal because of actually moving
    # the whole buffer in memory for every "shift" method call!
    while a.size > 0
      a.shift
    end
  }
}
```

== before the #568 fix (with 93e5d321e16ae0d84fedc7cc45478a8cad7f8730) ==
```
$ ./crystal run shiftperf.cr 
              user     system      total        real
shift     0.003799   0.000000   0.003799 (  2.495777)

$ ./crystal run --release shiftperf.cr 
              user     system      total        real
shift     0.003799   0.000000   0.003799 (  2.492066)

```
== after the #568 fix (with bf2912d56478154ddf8fa7a02c879ad1048bf57a) ==
```
$ ./crystal run shiftperf.cr 
              user     system      total        real
shift     0.066483   0.000000   0.066483 (  43.612190)

$ ./crystal run --release shiftperf.cr 
              user     system      total        real
shift     0.004654   0.000000   0.004654 (  3.056545)
```

Before the fix, shifting data in the array was using the SSE2 optimized memmove function from glibc, which is always faster than the one by one copy loop:
```
$ perf record ./shiftperf
# perf report
    99.44%  shiftperf  libc-2.20.so       [.] _wordcopy_fwd_dest_aligned      
     0.11%  shiftperf  libc-2.20.so       [.] __memmove_sse2                  
     0.05%  shiftperf  libgc.so.1.0.3     [.] 0x0000000000011c9e              
     0.04%  shiftperf  [kernel.kallsyms]  [k] debug_smp_processor_id          
     0.04%  shiftperf  libc-2.20.so       [.] __memcpy_sse2_unaligned         
     0.03%  shiftperf  [kernel.kallsyms]  [k] apic_timer_interrupt            
```

**Comment by ssvb** [comment]
On a side note, this benchmark runs very fast in Ruby. Which means a much better "unshift" implementation there.

**Comment by ssvb** [comment]
Renamed the issue, because it is in fact about Pointer#copy_from / Pointer#move_from performance (also sorry for the "shift"/"unshift" method names mixup). The potential Array#shift implementation improvements can be tracked in #573

**Comment by asterite** [comment]
This is fixed by https://github.com/manastech/crystal/commit/a2c3aef29e7b0a0f458559a6f05d035a0aafe20f

What I did was to use the intrinsics if the pointer classes were exactly the same, with a runtime check.

The thing is, if we have

```crystal
struct Pointer(T)
  def foo(x : Pointer(T))
  end
end

p1 = Pointer(Int32).null
p2 = Pointer(Int32 | String).null

p1.foo(p2) # compile error
p2.foo(p1) # ok
```

This is because in the first case foo's restriction becomes `Pointer(Int32)` and once the generic types match the restriction is applied to the type vars, and `Int32 | String` can't pass the "just" `Int32` restriction.

In the second case the restriction becomes `Pointer(Int32 | String)` and then `Int32` passes the `Int32 | String` restriction.

I first tried adding another overload to `move_from/copy_from` to use `self` as a restriction, but a `self` restriction just means "replace self with my type", which becomes `Pointer(Int32)` or `Pointer(Int32 | String)`, so it's exactly the same as the above reasoning. However, I'd like to have an overload for "exactly self", so maybe the current self restriction implementation is not correct.

On the other hand, the runtime check will always get optimized by llvm in release mode because the numbers will be hardcoded. And in the benchmarks you passed I didn't notice any difference between release/non-release mode with these changes. So for now this is OK I guess, I'll try to change the self restriction semantic later and see if things break.

Finally, Ruby's Array is a beast. They have a "shared" flag, so when you do shift/unshift/slice they actually return you an array whose pointer points to the array's buffer but in a different position. But this makes all mutable methods harder to implement, because then if you modified a shared copy you must create copies for the arrays the point to this one, etc. I don't think we'll ever do this. This is why I guess the performance in better in Ruby, but in Crystal we can implement other data structures if we need them, for example a Stack or a Queue.

However, it doesn't feel good to know that if you do `[1, 2, 3][1 .. -1].each { ... }` you are allocating an array just temporarily to iterate it, while in Ruby just a small allocation happens with a shifted pointer. In this case we can maybe have some other way to do it, for example obtaining an Iterator/Enumreator to a slice of the array, and then being able to iterate it, map it, etc. If you use the regular slice and find that it's your program bottleneck, you go and change it to the optimized way.
